### ISSUE TITLE
Going from browser console JS to scala-js-chrome code

### ISSUE THREAD

**Issue by phdoerfler** [AlexITC/scala-js-chrome#121]
# TL;DR

Browser JS Console:

```js
Object.entries(document.querySelector('.maplibregl-map'))
```
produces:
<img width="418" alt="image" src="https://user-images.githubusercontent.com/163428/158055210-f0c472b4-fbe1-447d-8895-a7b409c94f99.png">
👍

vs. the "same" in a chrome extension using ScalaJS:
```scala
import scala.scalajs.js
val evalResult = js.eval("Object.entries(document.querySelector('.maplibregl-map'))")
console.log(evalResult)
```
<img width="240" alt="image" src="https://user-images.githubusercontent.com/163428/158055258-7738b306-0a5c-4a77-a9f8-a0db9135b323.png">
👎

Why does this produce an empty array? How do I recreate what I did in Chrome's developer console?

---

## Investigation

fwiw, the produced JS is:
```js
$m_Lorg_scalajs_dom_package$().console__Lorg_scalajs_dom_raw_Console().log("eval:");
var evalResult = (0, eval)("Object.entries(document.querySelector('.maplibregl-map'))");
$m_Lorg_scalajs_dom_package$().console__Lorg_scalajs_dom_raw_Console().log(evalResult)
```

If I, manually, change the generated scalajs code to:

```js
$m_Lorg_scalajs_dom_package$().console__Lorg_scalajs_dom_raw_Console().log("eval:");
var evalResult = (0, eval)("Object.entries(document.querySelector('.maplibregl-map'))");
$m_Lorg_scalajs_dom_package$().console__Lorg_scalajs_dom_raw_Console().log(evalResult)
$m_Lorg_scalajs_dom_package$().console__Lorg_scalajs_dom_raw_Console().log(eval("Object.entries(document.querySelector('.maplibregl-map'))"))
$m_Lorg_scalajs_dom_package$().console__Lorg_scalajs_dom_raw_Console().log(Object.entries(document.querySelector('.maplibregl-map')))
```

I still get empty arrays:
<img width="228" alt="image" src="https://user-images.githubusercontent.com/163428/158055638-585929eb-0651-4d06-b47b-1ebc37f28d58.png">

It would appear as if JS run in the console has access to a wider scope, and indeed it has. It's called "top" and once I change it to the extension's scope (see [Chrome docs](https://developer.chrome.com/docs/devtools/console/reference/#context)):

<img width="351" alt="image" src="https://user-images.githubusercontent.com/163428/158055766-34f6ea71-b6d5-4a4b-9621-182735908a50.png">

then I, too, can no longer find those objects:

<img width="415" alt="image" src="https://user-images.githubusercontent.com/163428/158055781-5fca39c0-9715-41b1-9a3f-b94091f82f08.png">

It seems that, in order to escape from that extension or "content" context, I have to make use of [script injection](https://developer.chrome.com/docs/extensions/mv3/content_scripts/#functionality) aka content scripts.

---

## Where this is coming from

I am writing a Chrome extension adding functionality to the [route planner of komoot.com](https://komoot.com/plan). The extension has progressed quite a bit already but I want to take it a step further.
However, I am stuck.

I have arrived at a snippet of JS that, when entered in Chrome's JS console (via inspect element), gives me what I require:
`Object.entries(document.querySelector('.maplibregl-map')).find(([k, ]) => k.startsWith('__reactInternalInstance$'))[1].return.memoizedProps.value.map._listeners`:
<img width="1088" alt="image" src="https://user-images.githubusercontent.com/163428/158019407-dbb472fe-06cc-4591-ba56-6d5c8a5cf423.png">

I just can't figure out how to translate that to scala-js code using this wonderful sbt plugin.

My attempts so far:
```scala
import scala.scalajs.js.Object
def selected = document.querySelector(".maplibregl-map")
console.log("selected:")
console.log(selected)
```

results in:
<img width="900" alt="image" src="https://user-images.githubusercontent.com/163428/158019727-18b6dfc1-70d4-4e7a-a9f6-1cc4ad85bc3a.png">

just as I would had I entered that personally in the browser console:
<img width="912" alt="image" src="https://user-images.githubusercontent.com/163428/158019749-4e16f2bd-0dbe-4c15-ab64-a2d0396b87ac.png">

However, that is where the similarities end.

My Scala code (continued):
```scala
console.log(s"Object.entries of selected:")
console.log(Object.entries(selected))
```

logs this:
<img width="750" alt="image" src="https://user-images.githubusercontent.com/163428/158020980-ab318937-3e08-4b56-b51a-51adbcb48bc0.png">

To me, this looks more like a Scala object, likely the result of the scala-js compilation. This makes sense, of course, since `Object.entries` is essentially reflection in JS.

So I tried using `eval`:
```scala
console.log("eval:")
console.log(scala.scalajs.js.Dynamic.global.eval("Object.entries(document.querySelector('.maplibregl-map'))"))
```

But this gets me an empty array:
<img width="219" alt="image" src="https://user-images.githubusercontent.com/163428/158021140-68963e34-cf1d-40cf-b450-0809caa9eaf1.png">

Whereas in the browser, a `Object.entries(document.querySelector('.maplibregl-map'))` results in:
<img width="415" alt="image" src="https://user-images.githubusercontent.com/163428/158019819-1f85a5c8-5b60-417b-93db-53dc8cbb011a.png">

Again, this is likely something trivial and maybe the answer is staring right at me but I'm not seeing it. Thanks for any help!

**Comment by AlexITC** [comment]
I have linked this issue to the official scalajs chat, unfortunately, I won't have time to look into it in the coming days.

**Comment by phdoerfler** [comment]
I have updated the issue description with what I think is going on.
What I'm trying to say: Sorry for the spam @AlexITC, I'm working on this issue and it's looking like I can figure this out without additional help.
I am closing this issue for now. The reason being: Knowing what I know now, I would not have bothered to open it in the first place.
I'm still glad I documented it all here because it might help others but for that the issue doesn't have to remain open.
If you want to be kept in the loop feel free to reopen it. I do hope that I'll find time to work more on this next weekend.

In any case, thanks for all your work on this sbt-plugin without which none of this would have been possible!

**Comment by AlexITC** [comment]
Thanks for the details, if you liked the plugin, be sure to check https://github.com/AlexITC/chrome-scalajs-template

**Comment by phdoerfler** [comment]
Thanks for pointing out the template! I am, in fact, already using it. Saved me a ton of work and I'm not sure I'd have had the patience to piece these things together in my spare time.

In fact, code that shows how to execute JS on the "top" context so it can access the website's JS might be a nice additional thing to have in the template, don't you think? It's a common thing for extensions to, well, _extend_ a website with functionality. I shall file a PR when I get that to work!

**Comment by AlexITC** [comment]
> In fact, code that shows how to execute JS on the "top" context so it can access the website's JS might be a nice additional thing to have in the template, don't you think? It's a common thing for extensions to, well, extend a website with functionality. I shall file a PR when I get that to work!

Actually, I have that code, in short:
- The active tab context injects a js file to the target website like `<script>...</script>`.
- The injected script communicates to the active tab context through message passing `window.postMessage`
- The active tab listens for messages from the injected script, replying to those through message passing.

Let me see if I can prepare simple code about this.

**Comment by AlexITC** [comment]
@phdoerfler I have gathered most of the logic to do what you intend, there is a chance I missed some details because I didn't had time to test it, still, once the logic is fixed, you should be able to invoke `window.injected.getInfo()` that the extension has access to: https://github.com/AlexITC/chrome-scalajs-template/tree/website-injection

It's been a long time since I have been planning to include this in the template, I'm willing to fix it but I won't likely have much time in the following days.

Let me know how it goes!

**Comment by phdoerfler** [comment]
@AlexITC Alright, I finally had time to play around with it and it's looking great! I am actually able to access the website's JS. Thanks a lot!
I am a bit surprised by the sheer amount of code that you had to add in order to make this possible, but I can't argue with the results! I vote for merging this into the main branch asap.

**Comment by AlexITC** [comment]
Thanks for sharing, I'll try to run some manual tests and get this merged, good luck!

**Comment by AlexITC** [comment]
Thanks for letting me know, I merged the scala-steward's PR but never got back to add the securerandom library.

**Comment by AlexITC** [comment]
I pushed a fix to the linking error, also, I have merged the code injection to master (https://github.com/AlexITC/chrome-scalajs-template/commit/15227989f3ccdcaefdd5e3b40f58e265840371b5), to my surprise, it worked at the first try in Firefox/Chrome.

Thanks!
